File: Lintcode/149-Best-Time-To-Buy-And-Sell-Stock/BestTimeToBuyAndSellStock.java

public class BestTimeToBuyAndSellStock {
	/**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
	public int maxProfit(int[] prices) {
		int max = 0;
		if (prices == null || prices.length < 2) {
			return max;
		}
		int min = prices[0];
		for (int i = 1; i < prices.length; i++) {
			min = Math.min(min, prices[i]);
			max = Math.max(max, prices[i] - min);
		}
		return max;
	}
}

File: Leetcode/121-Best-Time-to-Buy-and-Sell-Stock/BestTimeToBuyAndSellStock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    if (!prices || prices.length < 2) {
        return 0;
    }
    var profit = 0
    var cost = prices[0]
    for (index in prices) {
        profit = Math.max(profit, prices[index]-cost)
        cost = Math.min(cost, prices[index])
    }
    return profit
};
